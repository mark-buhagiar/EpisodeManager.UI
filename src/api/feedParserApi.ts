import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import ParseHistoryListing from '../models/ParseHistoryListing';
import ParseHistoryDetails from '../models/ParseHistoryDetails';
import moment from 'moment';

const buildEndpoint = (endpoint: string): string => ApiConfig.feedParserApiBase + endpoint;

const feedParserMSEndpoints = {
    syncWithShowRss: buildEndpoint('sync/SyncWithShowRss'),
    forceSyncParseHistory: buildEndpoint('sync/ForceSyncParseHistory'),
    getParseHistoryBetweenDates: buildEndpoint('ParseHistory/GetParseHistoryBetweenDates'),
    getParseHistoryDetails: buildEndpoint('ParseHistory/GetParseHistoryDetails'),
};

export async function syncWithShowRss(): Promise<void> {
    try {
        await axios.post(feedParserMSEndpoints.syncWithShowRss);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

function convertDates(listing: ParseHistoryListing): ParseHistoryListing {
    return {
        ...listing,
        retrievedOn: moment.utc(listing.retrievedOn).toDate(),
        lastSyncedOn: moment.utc(listing.lastSyncedOn).toDate(),
    };
}

export async function getParseHistoryBetweenDates(startDate: Date, endDate: Date): Promise<ParseHistoryListing[]> {
    try {
        const params = {
            startDate,
            endDate,
        };
        const parseHistory = await axios.get<ParseHistoryListing[]>(feedParserMSEndpoints.getParseHistoryBetweenDates, {
            params,
        });

        return parseHistory.data.map(convertDates);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getParseHistoryDetails(id: number): Promise<ParseHistoryDetails> {
    try {
        const params = {
            parseHistoryId: id,
        };
        const parseHistoryDetails = await axios.get<ParseHistoryDetails>(feedParserMSEndpoints.getParseHistoryDetails, {
            params,
        });

        return parseHistoryDetails.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
export async function forceSyncParseHistory(id: number): Promise<void> {
    try {
        const params = {
            parseHistoryId: id,
        };

        await axios.post(feedParserMSEndpoints.forceSyncParseHistory, null, {
            params,
        });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
