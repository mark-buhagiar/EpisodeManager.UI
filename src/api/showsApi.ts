import axios from 'axios';
import Episode from '../models/Episode';
import ApiConfig from '../config/apiConfig';

const buildEndpoint = (endpoint: string): string => ApiConfig.showsApiBase + endpoint;

const showsMSEndpoints = {
    getForCurrentUserBetweenDates: buildEndpoint('episodes/getForCurrentUserBetweenDates'),
};

export async function getForCurrentUserBetweenDates(startDate: Date, endDate: Date): Promise<Episode[]> {
    try {
        const params = {
            startDate: startDate,
            endDate: endDate,
        };
        const result = await axios.get<Episode[]>(showsMSEndpoints.getForCurrentUserBetweenDates, { params });
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
