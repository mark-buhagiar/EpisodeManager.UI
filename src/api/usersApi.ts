import axios from 'axios';
import ApiConfig from '../config/apiConfig';
import EpisodePreferences from '../models/EpisodePreferences';
import Show from '../models/Show';

const buildEndpoint = (endpoint: string): string => ApiConfig.usersApiBase + endpoint;

const usersMSEndpoints = {
    addEpisodeDownloaded: buildEndpoint('subscriptions/AddEpisodeDownloaded'),
    getUserSubscriptionsForListing: buildEndpoint('subscriptions/GetUserSubscriptionsForListing'),
    unsubscribeFromShow: buildEndpoint('subscriptions/UnsubscribeFromShow'),
    subscribeToShow: buildEndpoint('subscriptions/SubscribeToShow'),
    setUserEpisodePreferences: buildEndpoint('users/setUserEpisodePreferences'),
    getUserEpisodePreferences: buildEndpoint('users/getUserEpisodePreferences'),
};

export async function addEpisodeDownloaded(episodeId: number): Promise<void> {
    try {
        const params = {
            episodeId,
        };
        await axios.put(usersMSEndpoints.addEpisodeDownloaded, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserSubscriptionsForListing(): Promise<Show[]> {
    try {
        const result = await axios.get(usersMSEndpoints.getUserSubscriptionsForListing);
        return result.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function unsubscribeFromShow(showId: number): Promise<void> {
    try {
        const params = {
            showId,
        };
        await axios.patch(usersMSEndpoints.unsubscribeFromShow, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function subscribeToShow(showIds: number[]): Promise<void> {
    try {
        const params = {
            showIds,
        };
        await axios.put(usersMSEndpoints.subscribeToShow, null, { params });
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function setUserEpisodePreferences(preferences: EpisodePreferences): Promise<void> {
    try {
        await axios.post(usersMSEndpoints.setUserEpisodePreferences, preferences);
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}

export async function getUserEpisodePreferences(): Promise<EpisodePreferences> {
    try {
        const response = await axios.get(usersMSEndpoints.getUserEpisodePreferences);
        return response.data;
    } catch (exception) {
        console.log(exception);
        throw exception;
    }
}
