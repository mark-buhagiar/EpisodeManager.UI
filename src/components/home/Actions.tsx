import React from 'react';
import ActionButton from './ActionButton';
import { useEpisodesSelectedListener } from '../../HoC/withEpisodesSelectedContext';
import './Button.scss';

interface Props {
    className: string;
}

const Actions: React.FC<Props> = ({ className }): JSX.Element => {
    const episodesSelected = useEpisodesSelectedListener();

    const component = (): JSX.Element => {
        return episodesSelected > 0 ? (
            <div className={`button-group ${className}`}>
                <ActionButton
                    dispatch={(null as unknown) as React.Dispatch<any>}
                    enabled={episodesSelected > 0}
                    label="Download"
                ></ActionButton>
                <ActionButton
                    dispatch={(null as unknown) as React.Dispatch<any>}
                    enabled={episodesSelected > 0}
                    label="Mark Watched"
                ></ActionButton>
            </div>
        ) : (
            <></>
        );
    };

    return component();
};

export default Actions;
