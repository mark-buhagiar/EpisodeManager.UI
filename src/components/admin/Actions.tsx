import React from 'react';
import Panel from '../common/panel';
import ButtonGroup from '../common/button/ButtonGroup';
import { ButtonProps } from '../common/button/Button';
import * as feedParserApi from '../../api/feedParserApi';
import './AdminActions.scss';

const AdminActions: React.FC = (): JSX.Element => {
    async function handleFetchNewEpisodesClicked(): Promise<void> {
        await feedParserApi.syncWithShowRss();
    }

    const buttons = [{ label: 'Fetch New Episodes', onClick: handleFetchNewEpisodesClicked }] as ButtonProps[];

    return (
        <Panel title="Actions">
            <ButtonGroup buttons={buttons} className="admin-actions" />
        </Panel>
    );
};

export default AdminActions;
