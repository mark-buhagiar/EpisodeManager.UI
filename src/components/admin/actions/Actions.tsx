import React, { useState } from 'react';
import Panel from '../../common/panel';
import ButtonGroup from '../../common/button/ButtonGroup';
import { ButtonProps } from '../../common/button/Button';
import * as feedParserApi from '../../../api/feedParserApi';
import './AdminActions.scss';

const AdminActions: React.FC = (): JSX.Element => {
    const [loading, setIsLoading] = useState(false);

    async function handleFetchNewEpisodesClicked(): Promise<void> {
        setIsLoading(true);
        await feedParserApi.syncWithShowRss();
        setIsLoading(false);
    }

    const buttons = [
        { label: 'Fetch New Episodes', onClick: handleFetchNewEpisodesClicked, enabled: !loading },
    ] as ButtonProps[];

    return (
        <Panel title="Actions">
            <ButtonGroup buttons={buttons} className="admin-actions" />
        </Panel>
    );
};

export default AdminActions;
