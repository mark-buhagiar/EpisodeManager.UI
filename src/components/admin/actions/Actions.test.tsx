import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import Actions from './Actions';
import { syncWithShowRss } from '../../../api/feedParserApi';

jest.mock('../../../api/feedParserApi', () => ({ syncWithShowRss: jest.fn() }));

describe('When clicking the fetch new episodes button', () => {
    beforeEach(() => {
        jest.useFakeTimers();
    });
    afterEach(() => {
        jest.useRealTimers();
        jest.resetAllMocks();
    });

    it('should render', (): void => {
        const { asFragment } = render(<Actions />);
        expect(asFragment()).toMatchSnapshot();
    });

    it(`should do nothing if successful`, async (): Promise<void> => {
        const { getByText } = render(<Actions />);
        (syncWithShowRss as jest.Mock).mockReturnValue(new Promise((resolve) => setTimeout(() => resolve(), 10)));

        const fetchButton = getByText('Fetch', { exact: false });
        fireEvent.click(fetchButton);

        expect(fetchButton.classList.contains('disabled')).toBeTruthy();
        expect(syncWithShowRss).toHaveBeenCalledTimes(1);
        act(() => {
            jest.advanceTimersByTime(10);
        });
        await act(async () => await Promise.resolve());
        expect(fetchButton.classList.contains('disabled')).toBeFalsy();
    });

    it(`should display an error when the request is unsuccessful`, async (): Promise<void> => {
        const { getByText } = render(<Actions />);
        (syncWithShowRss as jest.Mock).mockReturnValue(
            new Promise((resolve, reject) =>
                setTimeout(() => {
                    reject('Error test');
                }, 10),
            ),
        );

        const fetchButton = getByText('Fetch', { exact: false });
        fireEvent.click(fetchButton);

        expect(fetchButton.classList.contains('disabled')).toBeTruthy();
        expect(syncWithShowRss).toHaveBeenCalledTimes(1);
        act(() => {
            jest.advanceTimersByTime(10);
        });

        await act(async () => await Promise.resolve());
        expect(fetchButton.classList.contains('disabled')).toBeTruthy();
    });
});
