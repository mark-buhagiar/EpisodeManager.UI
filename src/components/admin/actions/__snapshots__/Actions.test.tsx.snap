// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When clicking the fetch new episodes button should render 1`] = `
<DocumentFragment>
  <div
    class="panel"
  >
    <div
      class="panel-title"
    >
      Actions
      <div
        class="panel-actions"
      />
      <div
        class="separator"
      />
    </div>
    <div
      class="button-group admin-actions"
    >
      <button
        class="button no-select "
        title=""
        type="button"
      >
        Fetch New Episodes
      </button>
    </div>
  </div>
</DocumentFragment>
`;
